### ISSUE TITLE
Pre-requisites & expectations for running Ghost-CLI

### ISSUE THREAD

**Issue by ErisDS** [TryGhost/Ghost-CLI#188]
I've been taking Ghost-CLI out for a spin in a production environment, and I've run into a few high-level conceptual questions about the high-level pre-requisites & expectations around running Ghost-CLI.

The TLDR; question here is how should your server be configured, what user & privileges should be in place, before installing & running ghost cli.

**Use Cases**

I think there are 3 main use cases for installing Ghost-CLI on a server, which need consideration, in order of importance:

1. Brand new blog on brand new install of Ubuntu (E.g. Fresh server)
2. Brand new blog on an existing install of Ubuntu (E.g. Adding Ghost to other software)
3. Switching from Ghost LTS to Ghost 1.0

We should absolutely optimise for case 1. This is the most common expectation for self-hosters. Case 2 should be rarer, but supported. Case 3 is temporary and so it doesn't need too much attention.

**Problem**

Pre-requisite steps are mostly a documentation topic, however there are three areas where I think there needs to be some attention to the expected flow as they impact on how the tool needs to work:

1. What configuration of user(s) should be used to install, and then run ghost-cli?
2. In what order should domain setup be done vs running the ssl setup, in order for that to work smoothly?
3. What configuration does MySQL require prior to running `ghost install`?

These 3 areas are where I tripped up attempting to get from a fresh install to fully working Ghost. There are a couple of related open issues to do with [sudo](https://github.com/TryGhost/Ghost-CLI/issues/164) and [ssl](https://github.com/TryGhost/Ghost-CLI/issues/176) and in particular the combo of questions 1 & 2 means I've not really managed to get ssl setup smoothly yet - I keep getting Exit code 1 and it's not possible to tell whether it's failing because of permissions or because the domain isn't reachable.

I have put together a [gist](https://gist.github.com/ErisDS/499af765e6a282632c5d13f6bf99d756) containing all the steps I eventually used to get a working Ghost install. Inside it are several notes which clarify where the confusion comes & what steps don't work in an obvious way.

I would like us to have a clear and consistent set of steps for the setup prior to running a `ghost` command in all 3 use cases, so that we can write this up in documentation.

I think we probably need to make some small adjustments for each of my 3 problem questions above, e.g. making sure ghost-cli can handle/expect the right level of user privileges & making it easily possible to run the ssl setup step as a detached second step if needs be.

**Comment by ErisDS** [comment]
Yo @sebgie this is the thing I was vaguely mumbling about the other day. I think your input on the high-level prerequisite steps here would be invaluable. 

@acburdine `ghost install` is working really well and I think having clarity around these 3 cases will really help to iron out the last few creases.

**Comment by acburdine** [comment]
Ok so this all makes sense, and here's my thoughts on each:

> 1. What configuration of user(s) should be used to install, and then run ghost-cli?

So I do think that the user running Ghost _should_ probably not have sudo permissions for security reasons, but unfortunately I don't think it's possible to do that for updating. The things we need sudo permissions for currently are:

1. Installing ghost-cli (if node/npm is installed via `apt`) - this probably only affects installing/updating ghost-cli itself, so this isn't as big of a deal necessarily
2. Setting up/starting/stopping systemd
3. Setting up nginx config files & restarting nginx
4. Creating the letsencrypt certificate - this is required because our letsencrypt setup stores all of the certs inside `/etc/letsencrypt`. This can be changed so that the certs are stored instead inside the ghost install directory, however that might cause issues with nginx itself, need to do more research to verify whether this is possible

Based on these sudo requirements, it should be possible to construct a user for ghost-cli that would only need sudo permissions for ubuntu's `service` command (used to interact with nginx and systemd). However, the initial `ghost install` command would need to be run with a user that has sudo permissions because ghost-cli needs to copy the systemd service and nginx config files into their proper locations. The actual `ghost start` command could however run as a different user.

---

> 2. In what order should domain setup be done vs running the ssl setup, in order for that to work smoothly?

Based on how letsencrypt works, the domain _does_ need to be setup before running the ssl setup, but fortunately the command to setup ssl is already detached. (see [this line](https://github.com/TryGhost/Ghost-CLI/blob/master/lib/services/nginx/index.js#L86) for the actual command.) It's not well documented because the method for defining custom commands inside a service at this point is still a hack (to be resolved hopefully by #179), but running `sudo ghost service nginx-ssl <email>` _should_ work on its own

My thought on how to fix this is to maybe output some additional help information _before_ running the ssl setup that says something along the lines of "Before running this please make sure you have the blog domain set up correctly". Maybe also add a ping check beforehand too that pings the url and if it doesn't resolve to the local ip (which node can find using [this](https://nodejs.org/api/os.html#os_os_networkinterfaces)) then error out?

---

> 3. What configuration does MySQL require prior to running ghost install?

This one should be pretty basic. Knex-Migrator can create the database if it does not already exist; however it's probably better in documentation to recommend creating the database first and giving specific, restricted access to a user just for Ghost.

Alternatively Ghost-CLI, using an account with elevated MySQL permissions (e.g. root), could run the raw SQL required to create the database and then add a specific ghost mysql user that only had permissions for that specific new database.

**Comment by sebgie** [comment]
> So I do think that the user running Ghost should probably not have sudo permissions for security reasons, but unfortunately I don't think it's possible to do that for updating.

Running Ghost as a system user is a requirement. `ghost-cli` can be used with sudo privileges to execute the tasks mentioned above. The systemd config should use the `User=` and `Group=` directives in the [Service] section to make sure that a user without password or shell is used.

I didn't check if there is an option to start Ghost with system call but if that's the case it would be required to switch users (e.g.: `su -u ghost node index.js`).

:+1: to everything else :-)

**Comment by ErisDS** [comment]
**On SSL**

> `sudo ghost service nginx-ssl <email>` _should_ work on its own

Does this mean that, if I'm trying to do it all in one go, that I should run `sudo ghost install`?

> It feels to me like SSL setup should be a separate step?

(this was @sebgie's comment on the gist I provided)

If we do this, how do we make sure that people know to do that step? Ideally, this step would be required. Again, the problem here is around making sure that real-world usecases work with the tooling we provide, and that the documentation / explanation of how to use the tools is absolutely clear so there can be no misunderstandings about how to go about using the tools.

---

**On MySQL**

> Knex-Migrator can create the database if it does not already exist; however it's probably better in documentation to recommend creating the database first and giving specific, restricted access to a user just for Ghost.

The flow with MySQL I find extremely confusing at the moment. There is no clarity about what Ghost-CLI expects, and I think this should be a one-clear-way to do it thing, not a we-kind-of-cope-no-matter-what thing.

How it works right now:

If you give ghost-cli root username & password, you can give it either an existing or non-existing database, and everything is fine either way. That's ok, but it seems weird to use the root user.

If you give ghost-cli a user you just created e.g. `CREATE USER ghost@localhost identified by '<ghost_mysql_pw>' and give it the ghost user and password, you must also create the database first and grant the ghost user privileges. Else, what happens is this:

```
dmin@rogueone:/opt/apps/test-ghost$ ghost setup
? Enter your blog URL: http://localhost:2368
? Enter your MySQL hostname: localhost
? Enter your MySQL username: ghost
? Enter your MySQL password: *******
? Enter your Ghost database name: ghost_test_prod
 ✔ System Stack
Your url contains a port. Skipping automatic nginx setup.
✔ Finishing setup
? Do you want to start Ghost? Yes
```

At this point it looks like everything is OK, because you've been asked if you want to start ghost, but then this happens...

```
 ✔ Validating config
An error occurred.
Message: 'ER_DBACCESS_DENIED_ERROR: Access denied for user 'ghost'@'localhost' to database 'ghost_test_prod''

Debug Information:
    Node Version: v6.10.2
    Ghost-CLI Version: 1.0.0-alpha.16
    Environment: production
    Command: 'ghost setup'
```

I think there should be _one clear way_ that this is done. E.g. you always give Ghost the root user, and it managed creating a less privileged user? Not sure, @sebgie I would like your input here on what an acceptable way to do this would be.

**Comment by acburdine** [comment]
So after a discussion between @sebgie and I, issues have been opened to address each of the three points mentioned above: #189, #190, and #191.

Will leave this open as an "epic" issue for this issue, and we can close this once those issues have been completed and documentation has been created to reflect those processes.

**Comment by acburdine** [comment]
All of the sub-issues have been finished, what remains is documentation

**Comment by sebgie** [comment]
This is a summary of the current state and which mechanisms `ghost-cli` follows when being installed:

**Enforced by ghost-cli:**
- a system user (`ghost`) is used for running systemd

The requirement to run with a system user makes it necessary to change the ownership of the `/content` directory. This also means that whenever a user needs to manually edit/add/delete a file in the `/content` directory they will need super user permissions (aka `sudo`). This could lead to the situation where single files or folders don't belong to the `ghost` user anymore. To fix this situation `ghost repair-permissions` and `ghost exec` (#317) will be implemented.

As it turned out during implementation the `ghost run` command is using the user that executes the cli command to run Ghost. This led to permission errors because the `/content` folder is owned by `ghost`. To overcome this problem the addition of a system user became mandatory for every Linux installation (#334).

- a new mysql user is generated if the mysql root user was provided

Adding a non-root user to mysql is done by ghost-cli (implemented in #191).

**Not enforced:**
- use a non-root user when using ghost-cli

This is highly recommended and all documentation assumes using a non-root user. The main difference when using a non-root user is that `sudo` is required to execute administrative tasks (install software, execute as another user...).

**Comment by acburdine** [comment]
Closing this as all of the smaller remaining pieces of this issue have been split out into separate, smaller issues.
